src/lib/services/transaction-service.ts

```diff
@@ -0,0 +1,370 @@
+      name: string; // Name of parameter
+      type: string; // Type, e.g., address, uint256
+      value: string; // Actual value
+    }>;
+  };
+  nonce?: number;
+  safeTxGas?: number;
+  safeTxHash?: string; // Generated transaction hash from safe contract
+  baseGas?: number;
+  confirmations?: Array<Confirmation>;
+  confirmationsRequired?: number;
+  gasPrice?: string;
+  executionDate?: string;
+  submissionDate?: string;
+  blockNumber?: number;
+  isExecuted?: boolean;
+  isSuccessful?: boolean;
+  origin?: string;
+}
+
+// Confirmation from transaction-service
+export interface Confirmation {
+  owner: string;
+  submissionDate: string;
+  transactionHash?: string;
+  signature: string;
+  signatureType: string;
+}
+
+export interface Safe {
+  // Safe from getSafeInfo
+  address: string;
+  nonce: number;
+  threshold: number;
+  owners: string[];
+  masterCopy: string;
+  modules: string[];
+  fallbackHandler: string;
+  guard: string;
+  version: string;
+  // serialized fields
+  isOrcaEnabled?: boolean;
+  isPod?: boolean;
+  hasToken?: boolean;
+}
+
+/**
+ * Decodes transaction given an ABI and the data and parses it to match Gnosis's dataDecoded field
+ * @param abi
+ * @param data
+ * @returns
+ */
+export function parseTransaction(
+  abi,
+  data: string,
+): {
+  method: string;
+  parameters: Array<{
+    name: string; // Name of parameter
+    type: string; // Type, e.g., address, uint256
+    value: string; // Actual value
+  }>;
+} {
+  const {
+    name,
+    args,
+    functionFragment: { inputs },
+  } = new ethers.utils.Interface(abi).parseTransaction({
+    data,
+  });
+
+  const parameters = inputs.map((input, index) => {
+    let value = args[index];
+    if (BigNumber.isBigNumber(value)) {
+      value = value.toString();
+    }
+    return {
+      name: input.name,
+      type: input.type,
+      value,
+    };
+  });
+  return {
+    method: name,
+    parameters,
+  };
+}
+
+/**
+ * Populates the "dataDecoded" field of a transaction, if one does not exist.
+ * @param transaction
+ */
+export async function populateDataDecoded(
+  safeTransaction: SafeTransaction,
+): Promise<SafeTransaction> {
+  // No data to decode, or was already provided the decoded data
+  if (
+    safeTransaction.dataDecoded ||
+    safeTransaction.data == null ||
+    safeTransaction.data === ethers.constants.HashZero
+  )
+    return safeTransaction;
+
+  const abi = await lookupContractAbi(safeTransaction.to);
+  const dataDecoded = abi ? parseTransaction(abi, safeTransaction.data) : null;
+  return { ...safeTransaction, dataDecoded };
+}
+
+/**
+ * Gets safe info from transaction service.
+ *
+ * @param address - Address of the Safe.
+ * @returns - Safe data.
+ */
+export async function getSafeInfo(address: string): Promise<Safe> {
+  // Transaction-service expects a checksum address, but graphql does not checksum its addresses
+  const result = await axios.get(`${gnosisUrl}/safes/${getAddress(address)}`);
```

Comment:
should we be checksumming the return val?

Reply:
We probably checksum the value down the line, and as long as the final product is checksummed I don't think it's that important.